k-distribution/src/main/scripts/lib/kit/constrained_term.py

```diff
@@ -0,0 +1,82 @@
+from dataclasses import dataclass
+from itertools import chain
+from typing import FrozenSet, Iterable, Optional, Tuple, Union
+
+from pyk.kast import KClaim, KInner, KRewrite, KRule, KVariable, flattenLabel
+from pyk.kastManip import (
+    pushDownRewrites,
+    splitConfigAndConstraints,
+    substitute,
+)
+from pyk.prelude import mlAnd
+
+from .utils import (
+    count_rhs_vars,
+    count_vars,
+    drop_var_prefixes,
+    extract_lhs,
+    extract_rhs,
+)
+
+
+@dataclass(frozen=True)
+class ConstrainedTerm:
+    config: KInner
+    constraints: FrozenSet[KInner]
+
+    def __init__(self, config: KInner, constraints: Iterable[KInner]) -> None:
+        constraints = frozenset(flattenLabel('#And', mlAnd(constraints)))
+        object.__setattr__(self, 'config', config)
+        object.__setattr__(self, 'constraints', constraints)
+
+    def __iter__(self):
+        return chain([self.config], self.constraints)
+
+    @staticmethod
+    def from_term(term) -> 'ConstrainedTerm':
+        config, constraint = splitConfigAndConstraints(term)
+        return ConstrainedTerm(config, [constraint])
+
+    @staticmethod
+    def from_claim(claim) -> Tuple['ConstrainedTerm', 'ConstrainedTerm']:
+        body = drop_var_prefixes(claim.body)
+        lhs = extract_lhs(body)
+        rhs = extract_rhs(body)
+        requires = drop_var_prefixes(claim.requires)
+        ensures = drop_var_prefixes(claim.ensures)
+        return ConstrainedTerm(lhs, [requires]), ConstrainedTerm(rhs, [ensures])
```

Comment:
I think splitting into LHS and RHS should be handled somewhere else perhaps, maybe as a method of `KInner`.

So this should instead return something like `ConstrainedTerm(claim.body, claim.requires #And claim.ensures)`.

We could also lift functions like `.lhs()` and `.rhs()` from `KInner` to this class as well, by computing, `removeUselessConstraints(mlAnd([getLHS(self.config)] + self.constraints))`.